types/block.go

```diff
@@ -318,10 +318,11 @@ func (b *Block) PutBlock(ctx context.Context, nodeAdder format.NodeAdder) error
 	// add namespaces to erasured shares and flatten the eds
 	leaves := flattenNamespacedEDS(namespacedShares, eds)
 
+	// create nmt adder wrapping batch adder
+	batchAdder := nodes.NewNmtNodeAdder(ctx, format.NewBatch(ctx, nodeAdder))
+
 	// iterate through each set of col and row leaves
 	for _, leafSet := range leaves {
-		// create a batch per each leafSet
-		batchAdder := nodes.NewNmtNodeAdder(ctx, format.NewBatch(ctx, nodeAdder))
```

Comment:
@evan-forbes, what was the reason to create a batch per `leafSet`? I was trying to find the answer myself, but I couldn't. Having one adder for all nodes seems like a better way to go for me.

Reply:
> what was the reason to create a batch per leafSet? 

each `leafSet` is all of the leaf data for a single row or column of the extended data square, so this is how we get each row and col root that go into the data availability header. While we will eventually write these batches asynchronously #204, this is needed using the current plugin/nmt structure.

Alternatively, I think we could write all of the data in a single batch using a different `NmtNodeAdder` structure, one that shares a batch for all each nmt.

Does that make any sense?

Reply:
I think the description of the `format.Batch` is misleading. You don't need to create a new `Batch` every time you have some group of nodes to be added. Instead, the `Batch` separates multiple additions into batches itself. So once a single `Batch` is created, you can use it extensively to add a lot of nodes.

Reply:
So regarding a TODO:

>// TODO: post data to IPFS in a goroutine

I don't think it is necessary as well, as `Add` and `AddMany` are almost never blocking for `Batch`.

Reply:
Once optionate Batch buffer [limit exceeds](https://github.com/ipfs/go-ipld-format/blob/master/batch.go#L136), it runs [async node addition](https://github.com/ipfs/go-ipld-format/blob/master/batch.go#L99). Also, it [only blocks](https://github.com/ipfs/go-ipld-format/blob/master/batch.go#L87) when there are `NumCpu` in-progress parallel batches.

Reply:
Does that make any sense?

Reply:
ahh I see now! You're right, that's awesome!